### DISCUSSION TITLE
Force yaml (single) value output to have (single) quotes

### DISCUSSION THREAD

**Discussion by StefMa** [apple/pkl#351] [Q&A] [answered]
I want to have the following output:
```yml
schedule:
- cron: '* /10 * * *'
- cron: '5 4 * * *'
```

This is just a `Listing<Mapping<String, String>>`.
```
schedule: Listing<Mapping<String, String>> = new {
  Map("cron", "*/10 * * * *").toMapping()
  Map("cron", "5 4 * * *").toMapping()
}
```
You can also check this out in the pkl-playground:
https://pkl-playground.vercel.app/?share=canal-throw-building

But unfortunately the output is the following:
```yml
schedule:
- cron: '*/10 * * *'
- cron: 5 4 * * *
```
**Notice the missing single quotes in the second `cron` list item**

It seems that the `YamlRenderer` determinate if they will wrap it into quotes depending on the first char.
For me it looks like the rule of "if its a special char use single quotes, otherwise not".

While this might be (I'm not 100% sure with that 😁 ) technically correct, I still want to force wrapping it into quotes.
I tested out different things* but don't get it to work.. Can someone help me here?! 🙏 

*
**Test cases**:
* Using `"'5 4 * * *'"` results in `'''5 4 * * *'''`
* Using `"\'5 4 * * *\'"` results in an error
* Using `"\\'5 4 * * *\\'"` results in `\'5 4 * * *\'`
* Wrapping into string literal (`"""`) produces the same outputs, so they don't change anything ☝️

**Comment by odenix** [comment]
YAML can be formatted in many different ways. I don’t think YamlRenderer gives users full control over the chosen syntax.

**Comment by bioball** [comment, marked as answer]
[Render directives](https://pkl-lang.org/package-docs/pkl/current/base/RenderDirective.html) are an escape hatch to get the renderer to do what you want.

Here's an example: this combines a render directive with an output converter to turn all strings into JSON strings, which is still valid YAML (YAML accepts JSON strings):

```groovy
cron = "* * * * *"

local jsonRenderer = new JsonRenderer {}

output {
  renderer = new YamlRenderer {
    converters {
      [String] = (it) -> new RenderDirective { text = " " + jsonRenderer.renderValue(it) }
    }
  }
}
```

This produces:

```yaml
cron: "* * * * *"
```

It'll be more work to produce strings with single quotes. If your goal is to have consistent looking strings, I'd just stick with double quotes here.

> While this might be (I'm not 100% sure with that 😁 ) technically correct

This is just a style issue; you can trust Pkl to produce the right thing here.

**Reply by StefMa** [reply]
Yay, this works 🎉 
Thank you @bioball 👏
